app/lib/search/text_utils.dart

```diff
@@ -111,3 +111,41 @@ Map<String, double> tokenize(String originalText) {
 }
 
 bool _isLower(String c) => c.toLowerCase() == c;
+
+/// Generates the N-grams of [input], which are continuous character strings of
+/// length between [minLength] and [maxLength] (both inclusive).
+/// Eg. abc -> ab, bc
+Set<String> ngrams(String input, int minLength, int maxLength) {
+  final ngrams = new Set<String>();
+  for (int length = minLength; length <= maxLength; length++) {
+    if (input.length > length) {
+      for (int i = 0; i <= input.length - length; i++) {
+        ngrams.add(input.substring(i, i + length));
+      }
+    }
+  }
+  return ngrams;
+}
+
+/// Generates lookup candidates that are, either:
+/// - derived by deleting one character from [input],
+/// - are the prefix part of [input] (min 4 characters)
+/// - are the suffix part of [input] (min 4 characters)
+Set<String> deriveLookupCandidates(String input) {
+  final set = new Set<String>();
```

Comment:
Should we rename input to `token` as the input is a token, right?

Also should we not include the `input` in `set` ?

Reply:
> Also should we not include the input in set ?

That could potentially add an extra, non-required entry in the `_lookupCandidates`, e.g. a token pointing to a Set that contains only that token. For the candidate calculation at lookup time we can cheaply add it to the appropriate set.